File: app/src/main/java/io/github/qwqgong/androidcyaml/AppControlService.java
package io.github.qwqgong.androidcyaml;

import android.app.Service;
import android.content.Intent;
import android.net.Uri;
import android.os.Binder;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.RemoteCallbackList;
import android.os.RemoteException;

public final class AppControlService extends Service implements RuntimeCoordinator.Listener {
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final RemoteCallbackList<IControlCallback> callbacks = new RemoteCallbackList<>();

    private RuntimeCoordinator coordinator;
    private RuntimeCoordinator.State state = RuntimeCoordinator.State.STOPPED;
    private String detail = "VPN 未连接";
    private String dashboardUrl = "";
    private int controllerPort;

    private final IAppControl.Stub binder = new IAppControl.Stub() {
        @Override
        public void registerCallback(IControlCallback callback) {
            enforceSameAppCaller();
            if (callback != null) {
                callbacks.register(callback);
                mainHandler.post(() -> sendSnapshot(callback));
            }
        }

        @Override
        public void unregisterCallback(IControlCallback callback) {
            enforceSameAppCaller();
            if (callback != null) {
                callbacks.unregister(callback);
            }
        }

        @Override
        public void restartRuntime(IOperationCallback callback) {
            enforceSameAppCaller();
            coordinator.restart((success, message) -> complete(callback, success, message));
        }

        @Override
        public void importConfig(Uri source, IOperationCallback callback) {
            enforceSameAppCaller();
            coordinator.importConfig(
                    source,
                    (success, message) -> complete(callback, success, message)
            );
        }
    };

    @Override
    public void onCreate() {
        super.onCreate();
        coordinator = RuntimeCoordinator.getInstance(this);
        coordinator.addListener(this);
    }

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public void onDestroy() {
        if (coordinator != null) {
            coordinator.removeListener(this);
        }
        callbacks.kill();
        super.onDestroy();
    }

    @Override
    public void onRuntimeStateChanged(
            RuntimeCoordinator.State newState,
            String newDetail,
            String newDashboardUrl,
            int newControllerPort
    ) {
        state = newState;
        detail = newDetail;
        dashboardUrl = newDashboardUrl;
        controllerPort = newControllerPort;
        broadcastSnapshot();
    }

    private void broadcastSnapshot() {
        int count = callbacks.beginBroadcast();
        try {
            for (int index = 0; index < count; index++) {
                sendSnapshot(callbacks.getBroadcastItem(index));
            }
        } finally {
            callbacks.finishBroadcast();
        }
    }

    private void sendSnapshot(IControlCallback callback) {
        try {
            callback.onStateChanged(
                    state.ordinal(),
                    detail,
                    AndroidVpnService.isAlwaysOnMode(),
                    AndroidVpnService.isLockdownMode(),
                    state == RuntimeCoordinator.State.RUNNING ? dashboardUrl : "",
                    state == RuntimeCoordinator.State.RUNNING ? controllerPort : 0
            );
        } catch (RemoteException ignored) {
            // RemoteCallbackList removes dead UI callbacks automatically.
        }
    }

    private static void complete(
            IOperationCallback callback,
            boolean success,
            String detail
    ) {
        if (callback == null) {
            return;
        }
        try {
            callback.onComplete(success, detail == null ? "" : detail);
        } catch (RemoteException ignored) {
            // The UI may have been reclaimed while the operation completed.
        }
    }

    private static void enforceSameAppCaller() {
        if (Binder.getCallingUid() != android.os.Process.myUid()) {
            throw new SecurityException("Only AndroidCyaml may use its control service");
        }
    }
}

File: app/src/main/java/io/github/qwqgong/androidcyaml/AppControlService.kt
package io.github.qwqgong.androidcyaml

import android.app.Service
import android.content.Intent
import android.net.Uri
import android.os.Binder
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.os.Process
import android.os.RemoteCallbackList
import android.os.RemoteException

class AppControlService : Service(), RuntimeStateBus.Listener {
    private val mainHandler = Handler(Looper.getMainLooper())
    private val callbacks = RemoteCallbackList<IControlCallback>()

    private lateinit var coordinator: RuntimeCoordinator
    private var snapshot: RuntimeSnapshot = RuntimeSnapshot.stopped()

    private val binder: IAppControl.Stub = object : IAppControl.Stub() {
        override fun registerCallback(callback: IControlCallback?) {
            enforceSameAppCaller()
            if (callback != null) {
                callbacks.register(callback)
                mainHandler.post { sendSnapshot(callback) }
            }
        }

        override fun unregisterCallback(callback: IControlCallback?) {
            enforceSameAppCaller()
            if (callback != null) {
                callbacks.unregister(callback)
            }
        }

        override fun restartRuntime(callback: IOperationCallback?) {
            enforceSameAppCaller()
            coordinator.restart { success, message -> complete(callback, success, message) }
        }

        override fun importConfig(source: Uri?, callback: IOperationCallback?) {
            enforceSameAppCaller()
            coordinator.importConfig(source) { success, message ->
                complete(callback, success, message)
            }
        }

        override fun getNetworkSelectionCatalog(callback: IOperationCallback?) {
            enforceSameAppCaller()
            coordinator.networkSelectionCatalog { success, message ->
                complete(callback, success, message)
            }
        }

        override fun setNetworkSelection(
            identity: String?,
            group: String?,
            target: String?,
            callback: IOperationCallback?,
        ) {
            enforceSameAppCaller()
            coordinator.setNetworkSelection(identity, group, target) { success, message ->
                complete(callback, success, message)
            }
        }

        override fun setRuntimeOverrides(
            processMatchingMode: String?,
            ipv6Enabled: Boolean,
            logLevel: String?,
            adaptiveTcpConcurrent: Boolean,
            webViewXhttp: Boolean,
            lanWebUiPublic: Boolean,
            callback: IOperationCallback?,
        ) {
            enforceSameAppCaller()
            coordinator.setRuntimeOverrides(
                processMatchingMode,
                ipv6Enabled,
                logLevel,
                adaptiveTcpConcurrent,
                webViewXhttp,
                lanWebUiPublic,
            ) { success, message -> complete(callback, success, message) }
        }
    }

    override fun onCreate() {
        super.onCreate()
        coordinator = RuntimeCoordinator.getInstance(this)
        coordinator.addListener(this)
    }

    override fun onBind(intent: Intent?): IBinder = binder

    override fun onDestroy() {
        if (this::coordinator.isInitialized) {
            coordinator.removeListener(this)
        }
        callbacks.kill()
        super.onDestroy()
    }

    override fun onRuntimeStateChanged(snapshot: RuntimeSnapshot) {
        this.snapshot = snapshot
        broadcastSnapshot()
    }

    private fun broadcastSnapshot() {
        val count = callbacks.beginBroadcast()
        try {
            for (index in 0 until count) {
                sendSnapshot(callbacks.getBroadcastItem(index))
            }
        } finally {
            callbacks.finishBroadcast()
        }
    }

    private fun sendSnapshot(callback: IControlCallback) {
        val current = snapshot
        val overrides = coordinator.runtimeOverrideSettings()
        try {
            callback.onStateChanged(
                RuntimeSnapshotPayload(
                    current.state.ordinal,
                    current.detail,
                    AndroidVpnService.isAlwaysOnMode(),
                    AndroidVpnService.isLockdownMode(),
                    if (current.state == RuntimeState.RUNNING) current.dashboardUrl else "",
                    if (current.state == RuntimeState.RUNNING) current.controllerPort else 0,
                    overrides.processMatchingMode.wireValue,
                    overrides.ipv6Enabled,
                    coordinator.effectiveIpv6Enabled(),
                    overrides.logLevel.wireValue,
                    overrides.adaptiveTcpConcurrent,
                    overrides.webViewXhttp,
                    overrides.lanWebUiPublic,
                ),
            )
        } catch (ignored: RemoteException) {
            // RemoteCallbackList removes dead UI callbacks automatically.
        }
    }

    private companion object {
        fun complete(callback: IOperationCallback?, success: Boolean, detail: String?) {
            if (callback == null) {
                return
            }
            try {
                callback.onComplete(success, detail ?: "")
            } catch (ignored: RemoteException) {
                // The UI may have been reclaimed while the operation completed.
            }
        }

        fun enforceSameAppCaller() {
            if (Binder.getCallingUid() != Process.myUid()) {
                throw SecurityException("Only AndroidCyaml may use its control service")
            }
        }
    }
}
